src/mono/mono/mini/intrinsics.c

```diff
@@ -614,6 +622,134 @@ MONO_RESTORE_WARNING
+		// managed implementation here as well.
+
+		MonoClass *tfrom_klass = mono_class_from_mono_type_internal (tfrom);
+		if (mono_class_is_nullable (tfrom_klass)) {
+			return NULL;
+		}
+
+		MonoClass *tto_klass = mono_class_from_mono_type_internal (tto);
+		if (mono_class_is_nullable (tto_klass)) {
+			return NULL;
+		}
+
+		// The same applies for when the type sizes do not match, as this will always throw
+		// and so its not an expected case and we can fallback to the managed implementation
+
+		int tfrom_align, tto_align;
+		gint32 size = mono_type_size (tfrom, &tfrom_align);
+
+		if (size != mono_type_size (tto, &tto_align)) {
+			return NULL;
+		}
+		g_assert (size < G_MAXUINT16);
+
+		// We have several different move opcodes to handle the data depending on the
+		// source and target types, so detect and optimize the most common ones falling
+		// back to what is effectively `ReadUnaligned<TTo>(ref As<TFrom, byte>(ref source))`
+		// for anything that can't be special cased as potentially zero-cost move.
+
+		guint32 opcode = OP_LDADDR;
+		MonoStackType tto_stack = STACK_OBJ;
+
+		bool tfrom_is_primitive_or_enum = false;
+		if (m_class_is_primitive(tfrom_klass)) {
+			tfrom_is_primitive_or_enum = true;
+		} else if (m_class_is_enumtype(tfrom_klass)) {
+			tfrom_is_primitive_or_enum = true;
+			tfrom_type = mono_class_enum_basetype_internal(tfrom_klass)->type;
+		}
+
+		bool tto_is_primitive_or_enum = false;
+		if (m_class_is_primitive(tto_klass)) {
+			tto_is_primitive_or_enum = true;
+		} else if (m_class_is_enumtype(tto_klass)) {
+			tto_is_primitive_or_enum = true;
+			tto_type = mono_class_enum_basetype_internal(tto_klass)->type;
+		}
+
+		if (tfrom_is_primitive_or_enum && tto_is_primitive_or_enum) {
+			if (size == 1) {
+				opcode = OP_MOVE;
+				tto_stack = STACK_I4;
+			} else if (size == 2) {
+				opcode = OP_MOVE;
+				tto_stack = STACK_I4;
+			} else if (size == 4) {
+				if ((tfrom_type == MONO_TYPE_R4) && ((tto_type == MONO_TYPE_I4) || (tto_type == MONO_TYPE_U4))) {
+					opcode = OP_MOVE_F_TO_I4;
+					tto_stack = STACK_I4;
+				} else if ((tto_type == MONO_TYPE_R4) && ((tfrom_type == MONO_TYPE_I4) || (tfrom_type == MONO_TYPE_U4))) {
+					opcode = OP_MOVE_I4_TO_F;
+					tto_stack = STACK_R4;
+				} else {
+					opcode = OP_MOVE;
+					tto_stack = STACK_I4;
+				}
+			} else if (size == 8) {
+				if ((tfrom_type == MONO_TYPE_R8) && ((tto_type == MONO_TYPE_I8) || (tto_type == MONO_TYPE_U8))) {
+					opcode = OP_MOVE_F_TO_I8;
+					tto_stack = STACK_I8;
+				} else if ((tto_type == MONO_TYPE_R8) && ((tfrom_type == MONO_TYPE_I8) || (tfrom_type == MONO_TYPE_U8))) {
+					opcode = OP_MOVE_I8_TO_F;
+					tto_stack = STACK_R8;
+				} else {
+					opcode = OP_MOVE;
+					tto_stack = STACK_I8;
+				}
+			}
+		} else if (mini_class_is_simd (cfg, tfrom_klass) && mini_class_is_simd (cfg, tto_klass)) {
+			opcode = OP_XMOVE;
+			tto_stack = STACK_VTYPE;
+		}
+
+		if (opcode == OP_LDADDR) {
```

Comment:
Was this the main thing that was missing from the pr I attempted it in? Looks to be the main new thing at a glance

What would cause such codegen given the actual managed signatures are all static methods taking value types?

Reply:
I only removed the assertion here. This test hit this code path:
https://github.com/dotnet/runtime/blob/37253e00c30e0ee234fa64a71222f9a7b0786ed1/src/libraries/System.Runtime/tests/System.Runtime.CompilerServices.Unsafe.Tests/UnsafeTests.cs#L1295-L1296